### ISSUE TITLE
[data grid] Document `onStateChange` prop

### ISSUE THREAD

**Issue by AndreSilva1993** [mui/mui-x#8582]
### Order ID or Support key 💳 (optional)

41693

### Duplicates

- [X] I have searched the existing issues

### Latest version

- [X] I have tested the latest version

### The problem in depth 🔍

Hi all! I've searched everywhere for my use case but didn't seem to find it nowhere and just want to be sure I'm not shooting myself in the foot. 

We have a lot of datagrids in our project and want to persist its state in local storage so that the user can resume its workflow when navigating back to the same page(s). However I feel that tapping into each individual method / event listener that might change relevant information that should be stored and calling the `exportState` there would become way too repetitive.

Would something like this be viable? I tested it and it works but I just want to be sure that memoizing / holding onto the `apiRef.current` to use it in the cleanup function does not bring any issues. Without doing this, in the effect's cleanup function the `apiRef.current` is already null.

```
export function usePersistedGridState(apiRef) {
  useEffect(() => {
    const gridRef = apiRef.current;

    return () => {
      console.log(gridRef.getAllColumns()); // save the columns visibility
      console.log(gridRef.exportState()); // save the grid's state
    };
  }, []);
}
```

### Your environment 🌎

<details>
  <summary>`npx @mui/envinfo`</summary>

```
`npx @mui/envinfo` hangs indefinitely for me but these are our versions:

"@mui/lab": "^5.0.0-alpha.104",
"@mui/material": "^5.10.10",
"@mui/styles": "^5.10.10",
"@mui/system": "^5.10.10",
"@mui/x-data-grid-pro": "^5.17.18",
```
</details>

**Comment by yaredtsy** [comment]
hey @AndreSilva1993 

cleanup functions are not guaranteed to be called in all scenarios. there is some condition it might not be called. so I do not think it's a good idea. 

>  I feel that tapping into each individual method / event listener that might change relevant information that should be stored and calling the exportState there would become way too repetitive.

`exportState` export the `state`, so you only need to subscribe to `stateChange` event.  and you do not have to use `getAllColumns` to save the column visibility. `getAllColumns` is just a selector function. it's selecting the columns from `state`. `exportState` is enough.

**Comment by AndreSilva1993** [comment]
Hi @yaredtsy, first of all thanks for the swift response. We are using the `onStateChange` event but the fact that it is invoked way too often (we're using a debouncing mechanism) and the fact that it is an undocumented event, it led me to believe that we didn't have the best approach to this.

Also the `getColumns` there was to circumvent this issue - https://github.com/mui/mui-x/issues/4620 - since we can't migrate to v6 yet and I was getting inconsistent exports from the `exportState` which does not always includes the `columnsVisibilityModel`.

So the advised approach in this scenario is to use the `onStateChange` then?

**Comment by yaredtsy** [comment]
> We are using the onStateChange event but the fact that it is invoked way too often (we're using a debouncing mechanism

yes sometimes it gets triggered unnecessarily, and also the state that is changed might not be exportable state so use deep equal, to check if the exportable state is changed. only then save the state.

```jsx
const prev = React.useRef();
 <DataGridPro
     ....
      onStateChange={(e) => {
          const current = apiRef.current.exportState();
          if (!_.isEqual(prev.current, current)) {
            prev.current = current;
            //save the state here
          }
        }}
 />
```

**Comment by yaredtsy** [comment]
> Also the getColumns there was to circumvent this issue - https://github.com/mui/mui-x/issues/4620 - since we can't migrate to v6 yet and I was getting inconsistent exports from the exportState which does not always include the columnsVisibilityModel.

have you tested it with latest version its seems to be working fine here : https://v5.mui.com/x/react-data-grid/state/#restore-the-state-with-initialstate

**Comment by AndreSilva1993** [comment]
@yaredtsy It does work indeed, we were on v5 a few PATCH versions behind. Ok will use the `onStateChange` with the deep equal comparison and since `exportState` returns everything I need, there's no need for other selectors. Feel free to mark this issue as solved and thank you so much for the help, it's really appreciated.

**Comment by AndreSilva1993** [comment]
@m4theushw that'd be awesome! I've seen some issues / comments mentioning this event so it would give developers more confidence in using it if it's documented. Thanks!

**Comment by jonathandewitt-dev** [comment]
While you're at it, why not add a feature that automatically persists the grid state for you? For example, I'm using the `id` prop as a key to save the grid state in local storage, but you could use IndexedDB or something. Would reduce quite a lot of boilerplate...

**Comment by clarson0** [comment]
built in state persistence would be a great feature, we've been having quite a bit of trouble with this, especially when columns change (which the documentation says needs to be stable), but there are cases where the columns have value getters that have dependencies. If there's an "easy" way to deal with that, i'd love to hear it

**Comment by michelengelen** [comment]
Thanks for the discussion. `onStateChange` is deliberately undocumented: it fires on every internal state update and exposes the raw internal state, which isn't a stable contract we want people building on. For state persistence, the supported path is `apiRef.current.exportState()` / `restoreState()` or the specific controlled callbacks: https://mui.com/x/react-data-grid/state/#save-and-restore-the-state.

Closing as not planned. If you have a use case that genuinely needs a documented catch-all event, please open a new issue describing it.
